clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
@ijuma got it. It would be little difficult to get such a test since the waiters object is private and we would require a reference to actual condition variable being inserted. Right now we can just get the size of the waiters queue. So what I am saying is that if we inserted 2 condition variables C1 and C2 in separate threads, it will pass regardless we do removeLast() or remove(object) since we can only check the size of the queue. It will require code change to Bufferpool.java to test the condition that you are hinting at.